app/src/main/java/org/mozilla/fenix/quickactionsheet/QuickActionSheet.kt

```diff
@@ -0,0 +1,125 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+   License, v. 2.0. If a copy of the MPL was not distributed with this
+   file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package org.mozilla.fenix.quickactionsheet
+
+import android.content.Context
+import android.util.AttributeSet
+import android.view.View
+import android.widget.LinearLayout
+import androidx.appcompat.widget.AppCompatImageButton
+import androidx.coordinatorlayout.widget.CoordinatorLayout
+import androidx.core.widget.NestedScrollView
+import com.google.android.material.bottomsheet.BottomSheetBehavior
+import mozilla.components.browser.toolbar.BrowserToolbar
+import org.mozilla.fenix.R
+import android.animation.ValueAnimator
+import android.view.accessibility.AccessibilityManager
+import androidx.interpolator.view.animation.FastOutSlowInInterpolator
+import org.mozilla.fenix.ext.increaseTapArea
+import org.mozilla.fenix.utils.Settings
+
+class QuickActionSheet @JvmOverloads constructor(
+    context: Context,
+    attrs: AttributeSet? = null,
+    defStyle: Int = 0,
+    defStyleRes: Int = 0
+) : LinearLayout(context, attrs, defStyle, defStyleRes) {
+
+    init {
+        inflate(getContext(), R.layout.layout_quick_action_sheet, this)
+    }
+
+    override fun onAttachedToWindow() {
+        super.onAttachedToWindow()
+        setupHandle()
+    }
+
+    private fun setupHandle() {
+        val handle = findViewById<AppCompatImageButton>(R.id.quick_action_sheet_handle)
+        val linearLayout = findViewById<LinearLayout>(R.id.quick_action_sheet)
+        val quickActionSheetBehavior = BottomSheetBehavior.from(linearLayout.parent as View) as QuickActionSheetBehavior
+
+        handle.increaseTapArea(grabHandleIncreasedTapArea)
+        handle.setOnClickListener {
+            bounceSheet(quickActionSheetBehavior)
+        }
+
+        val settings = Settings.getInstance(context)
+        if (settings.shouldAutoBounceQuickActionSheet) {
+            settings.incrementAutomaticBounceQuickActionSheetCount()
+            bounceSheet(quickActionSheetBehavior, demoBounceAnimationLength)
+        }
+    }
+
+    private fun bounceSheet(
+        quickActionSheetBehavior: QuickActionSheetBehavior,
+        duration: Long = bounceAnimationLength
+    ) {
+        val normalPeekHeight = quickActionSheetBehavior.peekHeight
+
+        val peakHeightMultiplier = if (duration == demoBounceAnimationLength)
+            demoBounceAnimationPeekHeightMultiplier else bounceAnimationPeekHeightMultiplier
+
+        val valueAnimator = ValueAnimator.ofFloat(normalPeekHeight.toFloat(),
+            normalPeekHeight * peakHeightMultiplier)
+
+        valueAnimator.addUpdateListener {
```

Comment:
It would be cleaner to have this many calls in a row scoped using a with {}, let {}, or run {} block, but this is okay. https://medium.com/@elye.project/mastering-kotlin-standard-functions-run-with-let-also-and-apply-9cd334b0ef84